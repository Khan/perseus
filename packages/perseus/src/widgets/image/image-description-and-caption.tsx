import * as React from "react";

import {PerseusI18nContext} from "../../components/i18n-context";
import Renderer from "../../renderer";

import styles from "./image-widget.module.css";

import type {APIOptions} from "../../types";
import type {PerseusImageBackground} from "@khanacademy/perseus-core";
import type {LinterContextProps} from "@khanacademy/perseus-linter";

interface Props {
    caption: string;
    backgroundImage: PerseusImageBackground;
    apiOptions: APIOptions;
    linterContext: LinterContextProps;
}

export const ImageDescriptionAndCaption = ({
    caption,
    backgroundImage,
    apiOptions,
    linterContext,
}: Props) => {
    const context = React.useContext(PerseusI18nContext);
    return (
        <div className={styles.descriptionAndCaptionContainer}>
            {/* Description code will go here */}

            {/* Caption */}
            {caption && (
                <figcaption
                    className="perseus-image-caption"
                    style={{
                        maxWidth: backgroundImage.width,
                    }}
                >
                    {/* The Renderer component is used here so that the caption
                        can support markdown and TeX. */}
                    <Renderer
                        content={caption}
                        apiOptions={apiOptions}
                        linterContext={linterContext}
                        strings={context.strings}
                    />
                </figcaption>
            )}
        </div>
    );
};
