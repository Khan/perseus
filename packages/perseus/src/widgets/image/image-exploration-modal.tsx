import {FlexibleDialog} from "@khanacademy/wonder-blocks-modal";
import {sizing} from "@khanacademy/wonder-blocks-tokens";
import {HeadingMedium} from "@khanacademy/wonder-blocks-typography";
import React from "react";

import AssetContext from "../../asset-context";
import {SvgImage} from "../../components";
import {PerseusI18nContext} from "../../components/i18n-context";
import Renderer from "../../renderer";

import styles from "./image-widget.module.css";

import type {Props as ImageProps} from "./image.class";

const MODAL_HEIGHT = 568;

interface Props extends ImageProps {
    longDescription: string; // required
}

export const ImageExplorationModal = (props: Props) => {
    const context = React.useContext(PerseusI18nContext);
    return (
        <div
            // We need to manually add the `framework-perseus` class so that
            // the modal can have the correct styling, even when the portal
            // makes it render outside the normal `framework-perseus` container.
            className={`framework-perseus ${styles.modalContainer}`}
        >
            <FlexibleDialog
                title={
                    <h1
                        className={`perseus-image-modal-title ${styles.modalTitleContainer}`}
                    >
                        {/* Use Renderer so that the title can support markdown and TeX. */}
                        <Renderer
                            content={
                                props.title
                                    ? props.title
                                    : context.strings.imageAlternativeTitle
                            }
                            apiOptions={props.apiOptions}
                            linterContext={props.linterContext}
                            strings={context.strings}
                        />
                    </h1>
                }
                content={<ImageExplorationModalContent {...props} />}
                styles={{
                    root: wbStyles.root,
                }}
            />
        </div>
    );
};

const ImageExplorationModalContent = (props: Props) => {
    const {
        backgroundImage,
        caption,
        alt,
        longDescription,
        linterContext,
        apiOptions,
        box,
        labels,
        range,
        trackInteraction,
    } = props;
    const context = React.useContext(PerseusI18nContext);

    if (!backgroundImage.height || !backgroundImage.width) {
        return null;
    }

    // Contain the image to the modal dimensions:
    // - Shrink image to the modal height if it's taller than the modal.
    // - Keep image its original size if it's shorter than the modal.
    // - Maintain the image's aspect ratio.
    const modalImageHeight = Math.min(MODAL_HEIGHT, backgroundImage.height);
    // bgWidth / bgHeight = X / modalImageHeight
    // => X = (bgWidth / bgHeight) * modalImageHeight
    const width =
        (backgroundImage.width / backgroundImage.height) * modalImageHeight;

    return (
        <div className={styles.modalPanelContainer}>
            <div className={styles.modalImageContainer}>
                {/* Need to use SvgImage in order to load Graphie images */}
                <AssetContext.Consumer>
                    {({setAssetStatus}) => (
                        <SvgImage
                            src={backgroundImage.url!}
                            alt={caption === alt ? "" : alt}
                            width={width}
                            height={modalImageHeight}
                            preloader={apiOptions.imagePreloader}
                            extraGraphie={{
                                box: box,
                                range: range,
                                labels: labels ?? [],
                            }}
                            trackInteraction={trackInteraction}
                            zoomToFullSizeOnMobile={apiOptions.isMobile}
                            constrainHeight={apiOptions.isMobile}
                            allowFullBleed={apiOptions.isMobile}
                            setAssetStatus={setAssetStatus}
                        />
                    )}
                </AssetContext.Consumer>
            </div>
            <div
                className={`perseus-image-modal-description ${styles.modalDescriptionContainer}`}
            >
                {caption && (
                    <div className={styles.modalCaptionContainer}>
                        {/* Use Renderer so that the caption can support markdown and TeX. */}
                        <Renderer
                            content={caption}
                            apiOptions={apiOptions}
                            linterContext={linterContext}
                            strings={context.strings}
                        />
                    </div>
                )}
                <HeadingMedium tag="h2" style={wbStyles.descriptionHeading}>
                    Description
                </HeadingMedium>
                {/* Use Renderer so that the description can support markdown and TeX. */}
                <Renderer
                    content={longDescription}
                    apiOptions={apiOptions}
                    linterContext={linterContext}
                    strings={context.strings}
                />
            </div>
        </div>
    );
};

const wbStyles = {
    root: {
        borderRadius: sizing.size_120,
        maxWidth: "100%",
    },
    descriptionHeading: {
        marginBlockEnd: sizing.size_160,
    },
};
