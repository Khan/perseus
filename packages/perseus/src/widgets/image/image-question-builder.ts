import type {
    PerseusRenderer,
    ImageWidget,
    PerseusImageBackground,
} from "@khanacademy/perseus-core";

export function imageQuestionBuilder(): ImageQuestionBuilder {
    return new ImageQuestionBuilder();
}

class ImageQuestionBuilder {
    private content: string = "[[â˜ƒ image 1]]";
    private backgroundImage: PerseusImageBackground = {};
    private title?: string;
    private caption?: string;
    private alt?: string;

    build(): PerseusRenderer {
        return {
            content: this.content,
            images: {},
            widgets: {
                "image 1": {
                    type: "image",
                    graded: true,
                    version: {major: 0, minor: 0},
                    static: false,
                    alignment: "default",
                    options: {
                        title: this.title,
                        caption: this.caption,
                        alt: this.alt,
                        backgroundImage: this.backgroundImage,
                    },
                } satisfies ImageWidget,
            },
        };
    }

    withContent(content: string): ImageQuestionBuilder {
        this.content = content;
        return this;
    }

    withImage(
        imageUrl: string,
        options?: {width?: number; height?: number},
    ): ImageQuestionBuilder {
        this.backgroundImage = {
            url: imageUrl,
            ...options,
        };
        return this;
    }

    withTitle(title: string): ImageQuestionBuilder {
        this.title = title;
        return this;
    }

    withCaption(caption: string): ImageQuestionBuilder {
        this.caption = caption;
        return this;
    }

    withAlt(alt: string): ImageQuestionBuilder {
        this.alt = alt;
        return this;
    }
}
