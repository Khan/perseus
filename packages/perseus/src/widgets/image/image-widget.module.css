.title-container {
    text-align: center;
    border-top: var(--wb-border-width-thin) solid
        var(--wb-semanticColor-core-border-neutral-subtle);
    padding-block-start: var(--wb-sizing-size_160);
    padding-block-end: var(--wb-sizing-size_160);
}

.description-and-caption-container {
    display: flex;
    flex-direction: row;
    margin-block-start: var(--wb-sizing-size_160);
    padding-block-end: var(--wb-sizing-size_160);
    border-bottom: var(--wb-border-width-thin) solid
        var(--wb-semanticColor-core-border-neutral-subtle);
}
