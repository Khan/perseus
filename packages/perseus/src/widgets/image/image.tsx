import {linterContextDefault} from "@khanacademy/perseus-linter";
import classNames from "classnames";
import * as React from "react";
import _ from "underscore";

import AssetContext from "../../asset-context";
import {PerseusI18nContext} from "../../components/i18n-context";
import SvgImage from "../../components/svg-image";
import * as Changeable from "../../mixins/changeable";
import Renderer from "../../renderer";
import noopValidator from "../__shared__/noop-validator";

import type {Range, PerseusImageWidgetOptions} from "../../perseus-types";
import type {ChangeFn, WidgetExports, WidgetProps, Widget} from "../../types";
import type {NullUserInput, PerseusImageRubric} from "../../validation.types";

const defaultBoxSize = 400;
const defaultRange: Range = [0, 10];
const defaultBackgroundImage = {
    url: null,
    width: 0,
    height: 0,
} as const;

const editorAlignments = ["block", "full-width"];

const DEFAULT_ALIGNMENT = "block";

type RenderProps = PerseusImageWidgetOptions; // there is no transform as part of exports

type ExternalProps = WidgetProps<RenderProps, PerseusImageRubric>;

type Props = ExternalProps & {
    alignment: NonNullable<ExternalProps["alignment"]>;
    title: NonNullable<ExternalProps["title"]>;
    range: NonNullable<ExternalProps["range"]>;
    box: NonNullable<ExternalProps["box"]>;
    backgroundImage: NonNullable<ExternalProps["backgroundImage"]>;
    labels: NonNullable<ExternalProps["labels"]>;
    alt: NonNullable<ExternalProps["alt"]>;
    caption: NonNullable<ExternalProps["caption"]>;
    linterContext: NonNullable<ExternalProps["linterContext"]>;
};

type DefaultProps = {
    alignment: Props["alignment"];
    title: Props["title"];
    range: Props["range"];
    box: Props["box"];
    backgroundImage: Props["backgroundImage"];
    labels: Props["labels"];
    alt: Props["alt"];
    caption: Props["caption"];
    linterContext: Props["linterContext"];
};

class ImageWidget extends React.Component<Props> implements Widget {
    static contextType = PerseusI18nContext;
    declare context: React.ContextType<typeof PerseusI18nContext>;

    static defaultProps: DefaultProps = {
        alignment: DEFAULT_ALIGNMENT,
        title: "",
        range: [defaultRange, defaultRange],
        box: [defaultBoxSize, defaultBoxSize],
        backgroundImage: defaultBackgroundImage,
        labels: [],
        alt: "",
        caption: "",
        linterContext: linterContextDefault,
    };

    // TODO (LEMS-2396): remove validation logic from widgets that don't validate
    static validate() {
        return noopValidator();
    }

    change: ChangeFn = (...args) => {
        return Changeable.change.apply(this, args);
    };

    getUserInput(): NullUserInput {
        return null;
    }

    // TODO (LEMS-2396): remove validation logic from widgets that don't validate
    simpleValidate() {
        return noopValidator();
    }

    render(): React.ReactNode {
        let image;
        const alt = this.props.caption === this.props.alt ? "" : this.props.alt;
        const {apiOptions} = this.props;

        const backgroundImage = this.props.backgroundImage;

        if (backgroundImage.url) {
            const url = backgroundImage.url;
            image = (
                <AssetContext.Consumer>
                    {({setAssetStatus}) => (
                        <SvgImage
                            src={url}
                            alt={alt}
                            width={backgroundImage.width}
                            height={backgroundImage.height}
                            preloader={apiOptions.imagePreloader}
                            extraGraphie={{
                                box: this.props.box,
                                range: this.props.range,
                                labels: this.props.labels,
                            }}
                            trackInteraction={this.props.trackInteraction}
                            zoomToFullSizeOnMobile={apiOptions.isMobile}
                            constrainHeight={apiOptions.isMobile}
                            allowFullBleed={apiOptions.isMobile}
                            setAssetStatus={setAssetStatus}
                        />
                    )}
                </AssetContext.Consumer>
            );
        }

        // For mobile we combine an image's title and caption.
        if (apiOptions.isMobile) {
            let titleAndCaption;

            if (this.props.title || this.props.caption) {
                let title = this.props.title;

                // Bold the title, and make it the first sentence of the
                // caption.
                if (title) {
                    // We add a period to separate the title from the caption
                    // (if it exists), unless the title already ends with a
                    // punctuation symbol (whitespace ignored). Copied from
                    // webapp: https://github.com/Khan/webapp/blob/6e930637edb65696d0749ea0f7558214aee32b4e/javascript/tutorial-shared-package/components/content-description.jsx#L80
                    // TODO(charlie): Internationalize this check, and the
                    // delimiter that is being inserted.
                    if (this.props.caption && !/[.?!"']\s*$/.test(title)) {
                        title += ".";
                    }

                    title = `**${title}** `;
                }

                const className = classNames({
                    "perseus-image-caption": true,
                    "has-title": !!title,
                });

                // Caption is left-aligned within a container that's centered
                // below the image, with these width constraints:
                //
                // 1. Size caption to width of the image on-screen.
                // 2. ... but constrain its width to a range based on the
                //    device to optimize readability - e.g. [320px, 450px] for
                //    phones.
                // 3. ... unless the image is floated, in which case we don't
                //    want the caption to overflow the image size.
                //
                // TODO(david): If caption is only 1 line long, center-align
                //     the text.
                const alignment = this.props.alignment;
                const isImageFullWidth =
                    alignment === "block" || alignment === "full-width";

                // This minWidth takes precedence over minWidth applied via
                // Aphrodite.
                const minWidth = isImageFullWidth ? null : "0 !important";

                titleAndCaption = (
                    <figcaption
                        className={className}
                        style={{
                            maxWidth: backgroundImage.width,
                        }}
                    >
                        <div
                            style={{
                                // @ts-expect-error - TS2322 - Type 'string | null' is not assignable to type 'MinWidth<string | number> | undefined'.
                                minWidth: minWidth,
                            }}
                        >
                            <Renderer
                                content={title + this.props.caption}
                                apiOptions={apiOptions}
                                linterContext={this.props.linterContext}
                                strings={this.context.strings}
                            />
                        </div>
                    </figcaption>
                );
            }

            return (
                <figure
                    className="perseus-image-widget"
                    style={{
                        maxWidth: backgroundImage.width,
                    }}
                >
                    {image}
                    {titleAndCaption}
                </figure>
            );
        }
        let title;
        let caption;

        if (this.props.title) {
            title = (
                <div className="perseus-image-title">
                    <Renderer
                        content={this.props.title}
                        apiOptions={apiOptions}
                        linterContext={this.props.linterContext}
                        strings={this.context.strings}
                    />
                </div>
            );
        }

        if (this.props.caption) {
            caption = (
                <figcaption
                    className="perseus-image-caption"
                    style={{
                        maxWidth: backgroundImage.width,
                    }}
                >
                    <Renderer
                        content={this.props.caption}
                        apiOptions={apiOptions}
                        linterContext={this.props.linterContext}
                        strings={this.context.strings}
                    />
                </figcaption>
            );
        }

        return (
            <figure
                className="perseus-image-widget"
                style={{
                    maxWidth: backgroundImage.width,
                }}
            >
                {title}
                {image}
                {caption}
            </figure>
        );
    }
}

export default {
    name: "image",
    // This widget's accessibility depends on its contents: if the image has
    // has a background but no alt text, it is not accessible
    accessible: (widgetOptions: PerseusImageWidgetOptions): boolean => {
        const bgImage = widgetOptions.backgroundImage;
        return !(bgImage && bgImage.url && !widgetOptions.alt);
    },
    defaultAlignment: DEFAULT_ALIGNMENT,
    supportedAlignments: editorAlignments,
    displayName: "Image",
    widget: ImageWidget,
    isLintable: true,
} as WidgetExports<typeof ImageWidget>;
