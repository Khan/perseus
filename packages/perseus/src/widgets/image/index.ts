export {default} from "./image.class";
