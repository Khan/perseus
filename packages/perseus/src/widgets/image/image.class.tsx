import {linterContextDefault} from "@khanacademy/perseus-linter";
import * as React from "react";
import _ from "underscore";

import {PerseusI18nContext} from "../../components/i18n-context";
import {getPromptJSON as _getPromptJSON} from "../../widget-ai-utils/image/image-ai-utils";

import {ImageComponent} from "./image";

import type {WidgetExports, WidgetProps, Widget} from "../../types";
import type {ImagePromptJSON} from "../../widget-ai-utils/image/image-ai-utils";
import type {Range, PerseusImageWidgetOptions} from "@khanacademy/perseus-core";

const defaultBoxSize = 400;
const defaultRange: Range = [0, 10];
const defaultBackgroundImage = {
    url: null,
    width: 0,
    height: 0,
} as const;

type ExternalProps = WidgetProps<PerseusImageWidgetOptions>;

export interface Props extends ExternalProps {
    alignment: NonNullable<ExternalProps["alignment"]>;
    title: NonNullable<ExternalProps["title"]>;
    range: NonNullable<ExternalProps["range"]>;
    box: NonNullable<ExternalProps["box"]>;
    backgroundImage: NonNullable<ExternalProps["backgroundImage"]>;
    labels: NonNullable<ExternalProps["labels"]>;
    alt: NonNullable<ExternalProps["alt"]>;
    caption: NonNullable<ExternalProps["caption"]>;
    linterContext: NonNullable<ExternalProps["linterContext"]>;
}

interface DefaultProps extends Partial<Props> {
    alignment: Props["alignment"];
    title: Props["title"];
    range: Props["range"];
    box: Props["box"];
    backgroundImage: Props["backgroundImage"];
    labels: Props["labels"];
    alt: Props["alt"];
    caption: Props["caption"];
    linterContext: Props["linterContext"];
}

class ImageWidget extends React.Component<Props> implements Widget {
    static contextType = PerseusI18nContext;
    declare context: React.ContextType<typeof PerseusI18nContext>;

    static defaultProps: DefaultProps = {
        alignment: "block",
        title: "",
        range: [defaultRange, defaultRange],
        box: [defaultBoxSize, defaultBoxSize],
        backgroundImage: defaultBackgroundImage,
        labels: [],
        alt: "",
        caption: "",
        linterContext: linterContextDefault,
    };

    // this just helps with TS weak typing when a Widget
    // doesn't implement any Widget methods
    isWidget = true as const;

    getPromptJSON(): ImagePromptJSON {
        return _getPromptJSON(this.props);
    }

    render(): React.ReactNode {
        return <ImageComponent {...this.props} />;
    }
}

export default {
    name: "image",
    displayName: "Image",
    widget: ImageWidget,
    isLintable: true,
} satisfies WidgetExports<typeof ImageWidget>;
