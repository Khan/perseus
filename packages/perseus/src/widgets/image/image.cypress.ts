import renderQuestionWithCypress from "../../../../../testing/render-question-with-cypress";
import {cypressTestDependencies} from "../../../../../testing/test-dependencies";
import * as Dependencies from "../../dependencies";
import * as Perseus from "../../index";

import {questionWithZoom} from "./image.testdata";

//  NOTE: The regression tests in this file use Cypress because they are intended to validate styling that is applied.
//        Since React Testing Library isn't applying the CSS to the elements,
//          we can't use Jest to verify that some keyboard interactions work properly.

describe("Image Widget", () => {
    beforeEach(() => {
        Dependencies.setDependencies(cypressTestDependencies);
        Perseus.init();
        window.innerWidth = 1024;
    });

    afterEach(() => {
        // remove classes that are added to the body
        cy.get("body").invoke("removeClass", "zoom-overlay-open");
        cy.get("img.zoom-img").invoke("remove");
    });

    it("opens and closes zoomable images on click", () => {
        // Arrange
        renderQuestionWithCypress(questionWithZoom);

        // Act - click on the image
        cy.get(".zoomable img").click();

        // Assert
        // The zoomed image should be visible and the zoomable image should be hidden
        cy.get("img.zoom-img").should("be.visible");
        cy.get(".zoomable img").should("not.be.visible");

        cy.get(".zoom-img").click();

        // Assert - the zoomable image should be hidden
        cy.get("img.zoom-img").should("not.be.visible");
    });

    it("opens and closes on keyboard interaction", () => {
        // Arrange
        renderQuestionWithCypress(questionWithZoom);

        // Act - focus on the zoomable image and press enter
        cy.get(".zoomable img").focus().type("{enter}");

        // Assert
        // The zoomed image should be visible and the zoomable image should be hidden
        cy.get("img.zoom-img").should("be.visible");
        cy.get(".zoomable img").should("not.be.visible");

        // Act - focus on the zoomed image and press escape
        cy.get("img.zoom-img").focus().type("{esc}");
    });
});
