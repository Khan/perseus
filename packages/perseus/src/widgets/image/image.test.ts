import {describe, beforeEach, it} from "@jest/globals";

import {testDependencies} from "../../../../../testing/test-dependencies";
import * as Dependencies from "../../dependencies";
import {scorePerseusItemTesting} from "../../util/test-utils";
import {isAccessible} from "../../widgets";
import {renderQuestion} from "../__testutils__/renderQuestion";

import {question} from "./image.testdata";

import type {APIOptions} from "../../types";

describe.each([true, false])("image widget - isMobile %b", (isMobile) => {
    const apiOptions: APIOptions = {isMobile};

    beforeEach(() => {
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("should snapshot", () => {
        // Arrange

        // Act
        const {container} = renderQuestion(question, apiOptions);

        // Assert
        expect(container).toMatchSnapshot("first render");
    });

    it("should be unanswerable", () => {
        // Arrange

        // Act
        const {renderer} = renderQuestion(question, apiOptions);
        const score = scorePerseusItemTesting(
            question,
            renderer.getUserInputMap(),
        );

        // Assert
        expect(score).toHaveBeenAnsweredIncorrectly();
    });

    it("should be accessible if background has 'alt' prop", () => {
        // Arrange, Act, and Assert
        expect(isAccessible(question.widgets["image 1"])).toBe(true);
    });

    it("should be inaccessible if background is missing 'alt' prop", () => {
        // Arrange
        const imageWidget = question.widgets["image 1"];
        const options = imageWidget.options;
        const inaccessibleWidgetInfo = {
            ...imageWidget,
            options: {
                ...options,
                alt: "",
            },
        };

        // Act and Assert
        expect(isAccessible(inaccessibleWidgetInfo)).toBe(false);
    });
});
