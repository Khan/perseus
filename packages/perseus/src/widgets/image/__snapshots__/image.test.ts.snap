// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`image widget - isMobile(false) should snapshot: first render 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive perseus-renderer-two-columns"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="perseus-two-columns"
      >
        <div
          class="perseus-column"
        >
          <div
            class="perseus-column-content"
          >
            <div
              class="paragraph"
            >
              <div
                class="perseus-widget-container widget-nohighlight widget-block"
              >
                <figure
                  class="perseus-image-widget"
                  style="max-width: 420px;"
                >
                  <div
                    class="perseus-image-title"
                  >
                    <div
                      class="perseus-renderer perseus-renderer-responsive"
                    >
                      <div
                        class="paragraph"
                        data-perseus-paragraph-index="0"
                      >
                        <div
                          class="paragraph"
                        >
                          Image Title
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="fixed-to-responsive svg-image"
                    style="max-width: 420px; max-height: 345px;"
                  >
                    <div
                      style="padding-bottom: 82.14%;"
                    />
                    <span
                      style="top: 0px; left: 0px; width: 100%; height: 100%; position: absolute; min-width: 20px; display: flex; justify-content: center; align-content: center;"
                    >
                      <div
                        class=""
                        style="align-items: stretch; border-width: 0px; border-style: solid; box-sizing: border-box; display: flex; flex-direction: column; margin: 0px; padding: 0px; position: relative; z-index: 0; min-height: 0; min-width: 0; justify-content: center;"
                      >
                        <svg
                          height="48"
                          viewBox="0 0 48 48"
                          width="48"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            class=""
                            d="M44.19 23.455a1.91 1.91 0 1 1 3.801 0h.003c.004.18.006.363.006.545 0 13.255-10.745 24-24 24S0 37.255 0 24 10.745 0 24 0c.182 0 .364.002.545.006V.01a1.91 1.91 0 1 1 0 3.801v.015A20.564 20.564 0 0 0 24 3.818C12.854 3.818 3.818 12.854 3.818 24c0 11.146 9.036 20.182 20.182 20.182 11.146 0 20.182-9.036 20.182-20.182 0-.182-.003-.364-.007-.545h.015z"
                            fill-rule="nonzero"
                            style="transform-origin: 50% 50%; animation-name: [object Object]; animation-duration: 1.1s; animation-iteration-count: infinite; animation-timing-function: linear; fill: rgba(33,36,44,0.50);"
                          />
                        </svg>
                      </div>
                    </span>
                  </div>
                  <figcaption
                    class="perseus-image-caption"
                    style="max-width: 420px;"
                  >
                    <div
                      class="perseus-renderer perseus-renderer-responsive"
                    >
                      <div
                        class="paragraph"
                        data-perseus-paragraph-index="0"
                      >
                        <div
                          class="paragraph"
                        >
                          Image Caption
                        </div>
                      </div>
                    </div>
                  </figcaption>
                </figure>
              </div>
            </div>
          </div>
        </div>
        <div
          class="perseus-column"
        >
          <div
            class="perseus-column-content"
          >
            <div
              class="paragraph"
            >
              A quilter wants to make the design shown at left using the Golden Ratio. Specifically, he wants the ratio of the triangle heights 
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  A:B
                </span>
                <span />
              </span>
               and 
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  B:C
                </span>
                <span />
              </span>
               to each equal 
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  1.62
                </span>
                <span />
              </span>
              . If the quilter makes the triangle height 
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  A=8\\ \\text{in}
                </span>
                <span />
              </span>
              , approximately how tall should he make triangle height 
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  C
                </span>
                <span />
              </span>
              ?
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`image widget - isMobile(true) should snapshot: first render 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive perseus-renderer-two-columns"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="perseus-two-columns"
      >
        <div
          class="perseus-column"
        >
          <div
            class="perseus-column-content"
          >
            <div
              class="paragraph"
            >
              <div
                class="perseus-widget-container widget-nohighlight widget-block"
              >
                <figure
                  class="perseus-image-widget"
                  style="max-width: 420px;"
                >
                  <div
                    class="fixed-to-responsive svg-image"
                    style="max-width: 420px; max-height: 345px;"
                  >
                    <div
                      style="padding-bottom: 82.14%;"
                    />
                    <span
                      style="top: 0px; left: 0px; width: 100%; height: 100%; position: absolute; min-width: 20px; display: flex; justify-content: center; align-content: center;"
                    >
                      <div
                        class=""
                        style="align-items: stretch; border-width: 0px; border-style: solid; box-sizing: border-box; display: flex; flex-direction: column; margin: 0px; padding: 0px; position: relative; z-index: 0; min-height: 0; min-width: 0; justify-content: center;"
                      >
                        <svg
                          height="48"
                          viewBox="0 0 48 48"
                          width="48"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            class=""
                            d="M44.19 23.455a1.91 1.91 0 1 1 3.801 0h.003c.004.18.006.363.006.545 0 13.255-10.745 24-24 24S0 37.255 0 24 10.745 0 24 0c.182 0 .364.002.545.006V.01a1.91 1.91 0 1 1 0 3.801v.015A20.564 20.564 0 0 0 24 3.818C12.854 3.818 3.818 12.854 3.818 24c0 11.146 9.036 20.182 20.182 20.182 11.146 0 20.182-9.036 20.182-20.182 0-.182-.003-.364-.007-.545h.015z"
                            fill-rule="nonzero"
                            style="transform-origin: 50% 50%; animation-name: [object Object]; animation-duration: 1.1s; animation-iteration-count: infinite; animation-timing-function: linear; fill: rgba(33,36,44,0.50);"
                          />
                        </svg>
                      </div>
                    </span>
                  </div>
                  <figcaption
                    class="perseus-image-caption has-title"
                    style="max-width: 420px;"
                  >
                    <div>
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <div
                            class="paragraph"
                          >
                            <strong>
                              Image Title.
                            </strong>
                             Image Caption
                          </div>
                        </div>
                      </div>
                    </div>
                  </figcaption>
                </figure>
              </div>
            </div>
          </div>
        </div>
        <div
          class="perseus-column"
        >
          <div
            class="perseus-column-content"
          >
            <div
              class="paragraph"
            >
              A quilter wants to make the design shown at left using the Golden Ratio. Specifically, he wants the ratio of the triangle heights 
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  A:B
                </span>
                <span />
              </span>
               and 
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  B:C
                </span>
                <span />
              </span>
               to each equal 
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  1.62
                </span>
                <span />
              </span>
              . If the quilter makes the triangle height 
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  A=8\\ \\text{in}
                </span>
                <span />
              </span>
              , approximately how tall should he make triangle height 
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  C
                </span>
                <span />
              </span>
              ?
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
